### DISCUSSION TITLE
Executing Hedy code step-per-per step

### DISCUSSION THREAD

**Discussion by redelaar** [hedyorg/hedy#2483] [Implementation discussions]
Hi @Felienne , @OnnoEbbens  and others!

For my bachelor's project, I am working on implementing new debugging tools into Hedy. A handy feature according to @OnnoEbbens would be to execute Hedy line per line. In this discussion, I want to share the design ideas that I have so far. 

I am not sure if this idea can be executed during my project due to time constraints. However, I still think it would be very cool project! I wanted to share the designs to get some Feedback and maybe a new batch of bachelor students will implement this, or someone else, or maybe me.

### Designs made so far (made with paint/photoshop): 3 images

The executing of lines step-per-step would work together with the newly created variable view (from this [PR: click for a short video](https://github.com/Felienne/hedy/pull/2241)).

**1. In the normal view, a new button is present. Like many IDEs, there's a button for running and debugging**
![1](https://user-images.githubusercontent.com/33595265/162500476-d52a02f8-77de-4c07-b589-9ba5b5d4bbd5.png)

**2. When the 🐞 debugging button has been clicked, a step-per-step program appears. Three new buttons appear: next line, restart and stop. The line that is being focused on has a blue background (#2D6099)**
![2](https://user-images.githubusercontent.com/33595265/162500007-2ab9e490-8b50-46df-9b24-64e88fcf5ac5.png)

**3. Just like in regular IDEs, the highlighted line will be executed AFTER clicking the "⏩" button.** 
![3](https://user-images.githubusercontent.com/33595265/162500011-0fd93d23-1a19-46aa-b9f6-1f57a07e7fc4.png)

Please tell me if I misplaced this discussion and should place it somewhere else.

**Comment by TiBiBa** [comment]
I really like this idea! I do have some concerns as where to implement this feature, as with the current overload of adventures and functionalities it gives me a bit of a "kid in a candy store" feeling. Too many options to choice from that users lose the overview and have no clue where to start programming.

Maybe can implement this as a separate section within Hedy? There is also a student currently working on implementation Parsons problem into the Adventures. Also a really nice feature, but I'm not really sure if we should keep that in adventures or make it a separate flow children can follow. It might be nice to re-think our current UI and do a complete re-styling so we can better support new features like this. It might also be nice for teachers to enable/disable the debugging mode. 

TLDR; I really like this idea! But we do have to think about the UI implementation to prevent the interface from getting too crowded.

**Reply by redelaar** [reply]
Thanks for the input Timon! I've spent some time creating a POC to see whether the idea is possible (functionality-wise), and it is! So that is nice. 

When it comes to the implementation, I'm not sure what the best move would be. I agree with your suggestions such as abling/disabling it by teachers!
